runtime/intern.c

```diff
@@ -177,6 +183,21 @@ static uintnat read64u(struct caml_intern_state* s)
 }
 #endif
 
+static int readvlq(struct caml_intern_state* s, /*out*/ uintnat * res)
+{
+  unsigned char c = read8u(s);
+  uintnat n = c & 0x7F;
+  int retcode = 0;
+  while ((c & 0x80) != 0) {
+    c = read8u(s);
+    uintnat n7 = n << 7;
+    if (n != n7 >> 7) retcode = -1;
```

Comment:
I've checked that this works correctly under '-fsanitize-undefined' (unsigned integer overflow is meant to be well defined, and the code here is meant to check for overflow when writing out a VLQ on 64-bit platform and reading it back on 32).